master/internal/stream/projects.go

```diff
@@ -0,0 +1,226 @@
+package stream
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"github.com/uptrace/bun"
+
+	"github.com/determined-ai/determined/master/internal/db"
+	"github.com/determined-ai/determined/master/pkg/model"
+	"github.com/determined-ai/determined/master/pkg/stream"
+)
+
+const (
+	// ProjectsDeleteKey specifies the key for delete projects.
+	ProjectsDeleteKey = "projects_deleted"
+	// ProjectsUpsertKey specifies the key for upsert projects.
+	ProjectsUpsertKey = "project"
+)
+
+// ProjectMsg is a stream.Msg.
+type ProjectMsg struct {
```

Comment:
note: look into the comparison between using model.Project (and similar structs as baselines for these messages to make sure we don't diverge from existing patterns)

Reply:
ProjectMsg
Pros:
- Decouples streaming from master 
- Streaming Updates owns the definitions for what streamable types are 

Cons:
- Introduces a new requirement to manually update streaming updates whenever there is a change to the data model to include new field (is this a con, or the intended behavior)

model.Project
Pros:
- Ensures that streaming updates' entities are updates automatically with any model changes

Cons:
- Additional fields that are not included in the original table would need to be omitted
- Any changes to model structs will also break any clients built on top of streaming updates if they are not flexible enough to adjust to new definitions (shouldn't be an issue if communication is being done over json objects & the client is not rigid of it's definitions or shares the same definition as model)

There is an argument that could be made for requiring any data model changes of this scale to also require streaming updates to be updated; however, knowing when this change should happen could be tricky if a row is being added. We might want to have that extra step put in place to ensure that we are certain that we want to add the additional row, but it could easily be forgotten. The same could be said of the inverse, of us not considering that we didn't want that information to be shared over streaming & in the event that model struct definition is used and then we might not catch the new information being shared.

Reply:
I'm a proponent for having separate specialized structs for streaming updates, with the caveat that we introduce a mechanism for notifying folks when a change is modifying the definition of a table that is related to a streamable entity, that we confirm that we either want/don't want that information included in streamings; this could be as simple as having a test that compares the columns in the DB to the field in the Msg struct & making sure the expected fields are present/absent

Reply:
I also like the specialized structs.

Remember that one of the parts of our REST API that makes it inefficient is the fact that the proto objects (like the model.* objects) include lots of information originating from several tables.

And explicit design goal of streaming updates was "no joining of data between tables... if the client wants data from multiple tables they'll have to stream it and assemble it themselves".